tests/unit/test_remote_configuration.py

```diff
@@ -0,0 +1,332 @@
+    ConfigReadError,
+    RemoteConfigurationProvider,
+    load_config_file,
+)
+from deepdiff import DeepDiff  # type: ignore[import]
+from helpers import k8s_resource_multipatch
+from ops import testing
+from ops.charm import CharmBase, CharmEvents
+from ops.framework import EventBase, EventSource
+from ops.model import BlockedStatus
+
+from charm import AlertmanagerCharm
+
+logger = logging.getLogger(__name__)
+
+testing.SIMULATE_CAN_CONNECT = True
+
+TEST_APP_NAME = "provider-tester"
+METADATA = f"""
+name: {TEST_APP_NAME}
+provides:
+  {RELATION_NAME}:
+    interface: alertmanager_remote_configuration
+"""
+TEST_ALERTMANAGER_CONFIG_WITHOUT_TEMPLATES_FILE_PATH = "./tests/unit/test_config/alertmanager.yml"
+TEST_ALERTMANAGER_CONFIG_WITH_TEMPLATES_FILE_PATH = (
+    "./tests/unit/test_config/alertmanager_with_templates.yml"
+)
+TEST_ALERTMANAGER_INVALID_CONFIG_FILE_PATH = "./tests/unit/test_config/alertmanager_invalid.yml"
+TEST_ALERTMANAGER_TEMPLATES_FILE_PATH = "./tests/unit/test_config/test_templates.tmpl"
+
+
+class AlertmanagerConfigFileChangedEvent(EventBase):
+    pass
+
+
+class AlertmanagerConfigFileChangedCharmEvents(CharmEvents):
+    alertmanager_config_file_changed = EventSource(AlertmanagerConfigFileChangedEvent)
+
+
+class RemoteConfigurationProviderCharm(CharmBase):
+    ALERTMANAGER_CONFIG_FILE = TEST_ALERTMANAGER_CONFIG_WITHOUT_TEMPLATES_FILE_PATH
+
+    on = AlertmanagerConfigFileChangedCharmEvents()
+
+    def __init__(self, *args):
+        super().__init__(*args)
+
+        try:
+            alertmanager_config = load_config_file(self.ALERTMANAGER_CONFIG_FILE)
+        except ConfigReadError:
+            alertmanager_config = {}
+        self.remote_configuration_provider = RemoteConfigurationProvider(
+            charm=self,
+            alertmanager_config=alertmanager_config,
+            relation_name=RELATION_NAME,
+        )
+
+        self.framework.observe(self.on.alertmanager_config_file_changed, self._update_config)
+
+    def _update_config(self, _):
+        try:
+            relation = self.model.get_relation(RELATION_NAME)
+            try:
+                alertmanager_config = load_config_file(self.ALERTMANAGER_CONFIG_FILE)
+            except ConfigReadError:
+                alertmanager_config = {}
+            self.remote_configuration_provider.update_relation_data_bag(
+                alertmanager_config, relation
+            )
+        except ConfigReadError:
+            logger.warning("Error reading Alertmanager config file.")
+
+
+class TestAlertmanagerRemoteConfigurationProvider(unittest.TestCase):
+    def setUp(self) -> None:
+        self.harness = testing.Harness(RemoteConfigurationProviderCharm, meta=METADATA)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.set_leader(True)
+        self.harness.begin_with_initial_hooks()
+
+    def test_config_without_templates_updates_only_alertmanager_config_in_the_data_bag(self):
```

Comment:
Technically, I'm not trying to test alertmanager here, but the library. And I believe we do want to test the library. 
Also `provider` side, doesn't emit any events, it's just loading the data to the relation data bag.
One thing I realized after this comment, is that I shouldn't test for non existing file, because this part is actually controlled by the provider charm. I should test for empty config instead.

Reply:
Since the library is also going here, the tests do make sense here. It may be more logical to split them into `tests/unit/test_remote_configuration_[provider|consumer].py` so it's obvious what each one does and the boilerplate for testing is as isolated and minimal as necessary, though

Reply:
Done.